File: routes/transactionRoutes.js
/**
 * Transaction Routes
 * 
 * Defines routes for handling transactions within the application. This includes
 * creating new transactions, fetching transactions for all users or a specific user,
 * and marking transactions as returned. The routes utilize the TransactionService
 * to interact with the database and perform the necessary operations.
 * 
 * Routes:
 * POST / - Create a new transaction
 * GET / - Get all transactions
 * GET /user/:userId - Get all transactions for a specific user
 * PATCH /:id/return - Mark a transaction as returned
 * PUT /:id - Updates a transaction by its ID
 * DELETE /:id - Deletes a transaction by its ID
 * 
 * @version 1.0
 * @author Hasan Siddiqui
 */

const express = require('express');
const router = express.Router();
const path = require('path');

// Universal path to the TransactionService
const TransactionService = require(path.join(__dirname, '..', 'services', 'TransactionService'));

// Route to create a new transaction
router.post('/', async (req, res) => {
  try {
    const newTransaction = await TransactionService.createTransaction(req.body);
    res.status(201).json(newTransaction);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
});

// Route to get all transations
router.get('/', async (req, res) => {
    try {
      const transactions = await TransactionService.getAllTransactions();
      res.json(transactions);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  });

// Route to get all transactions for a user
router.get('/user/:userId', async (req, res) => {
  try {
    const transactions = await TransactionService.getTransactionsByUser(req.params.userId);
    res.json(transactions);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Route to mark a transaction as returned
router.patch('/:id/return', async (req, res) => {
  try {
    const returnedTransaction = await TransactionService.markAsReturned(req.params.id);
    if (returnedTransaction) {
      res.json(returnedTransaction);
    } else {
      res.status(404).json({ message: 'Transaction not found' });
    }
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Route to update a book by id
router.put('/:id', async (req, res) => {
    try {
      const updatedTransaction = await TransactionService.updateTransaction(req.params.id, req.body);
      if (updatedTransaction) {
        res.json(updatedTransaction);
      } else {
        res.status(404).json({ message: 'Transaction not found' });
      }
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  });

// Route to delete a book by id
router.delete('/:id', async (req, res) => {
    try {
      const transactionDeleted = await TransactionService.deleteTransaction(req.params.id);
      if (transactionDeleted) {
        res.json({ message: 'Transaction deleted successfully' });
      } else {
        res.status(404).json({ message: 'Transaction not found' });
      }
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  });

module.exports = router;

File: routes/transaction_routes.py
from flask import Blueprint, request, jsonify
from models.transaction import Transaction  # Adjust import path as necessary
from mongoengine.errors import ValidationError, DoesNotExist

transaction_routes = Blueprint('transaction_routes', __name__)

# Route to create a new transaction
@transaction_routes.route('/', methods=['POST'])
def create_transaction():
    try:
        transaction_data = request.json
        transaction = Transaction(**transaction_data).save()
        return jsonify(transaction), 201
    except ValidationError as e:
        return jsonify({'error': str(e)}), 400

# Route to get all transactions
@transaction_routes.route('/', methods=['GET'])
def get_all_transactions():
    transactions = Transaction.objects()
    return jsonify(transactions), 200

# Route to get all transactions for a specific user
@transaction_routes.route('/user/<user_id>', methods=['GET'])
def get_user_transactions(user_id):
    transactions = Transaction.objects(userId=user_id)
    return jsonify(transactions), 200

# Route to mark a transaction as returned
@transaction_routes.route('/<transaction_id>/return', methods=['PATCH'])
def mark_transaction_returned(transaction_id):
    try:
        transaction = Transaction.objects.get(id=transaction_id)
        transaction.returned = True
        transaction.save()
        return jsonify(transaction), 200
    except DoesNotExist:
        return jsonify({'error': 'Transaction not found'}), 404

# Route to update a transaction by ID
@transaction_routes.route('/<transaction_id>', methods=['PUT'])
def update_transaction(transaction_id):
    try:
        transaction_data = request.json
        transaction = Transaction.objects.get(id=transaction_id)
        transaction.update(**transaction_data)
        return jsonify(transaction.reload()), 200
    except (DoesNotExist, ValidationError) as e:
        return jsonify({'error': str(e)}), 400

# Route to delete a transaction by ID
@transaction_routes.route('/<transaction_id>', methods=['DELETE'])
def delete_transaction(transaction_id):
    try:
        transaction = Transaction.objects.get(id=transaction_id)
        transaction.delete()
        return jsonify({'message': 'Transaction deleted successfully'}), 200
    except DoesNotExist:
        return jsonify({'error': 'Transaction not found'}), 404
